### PR TITLE
feature: add digital signature attributes

### PR DESCRIPTION
**Pull request by Darbiol**
- add enable/disable digital signature on templates
- add digital signature attribute to hold observee/observer info

[S-08693](https://www12.v1host.com/SchoolImprovementNetwork47/story.mvc/Summary?oidToken=Story:169233) and [S-08879](https://www12.v1host.com/SchoolImprovementNetwork47/story.mvc/Summary?oidToken=Story:177578)

### CODE REVIEW

**Comment by Darbiol** [comment]
@mashu-daishi @MrAnderson0309 @hanchan07 can you please take a look at this PR, see if you're OK with this

File: schemas/observation/observation.js [hunk 1]
```diff
@@ -102,7 +102,14 @@ module.exports = function ( level ) {
 		 * CONCERNS
 		 * Where did this come from? Its not in the wiki
 		*/
-		'manualScore' : Joi.number().optional().description( 'The manually-assigned score of the observation' )
+		'manualScore' : Joi.number().optional().description( 'The manually-assigned score of the observation' ),
+
+		// Digital signature attribute
+		'signature' : Joi
```
Comment (mashu-daishi) [inline schemas/observation/observation.js]: This should probably be keyed the same as the index.js file (it's listed as `digitalSignature` there).

File: schemas/observation/digital-signature.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	'isValid' : Joi.boolean().optional().description( 'flag that determines if the signature is valid or not' )
+};
+
+module.exports = {
+	'observer' : Joi
+		.array()
+		.optional()
+		.items( Joi
+			.object()
+			.optional()
```
Comment (mashu-daishi) [inline schemas/observation/digital-signature.js:17]: Are you allowing for empty arrays within the `observer` and `observee` fields? I believe listing the items as `optional()` allows for that.

File: schemas/observation/digital-signature.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+			.keys( signatureInfo ) )
+		.description( 'digital signature info for observer' ),
+
+	'observee' : Joi
```
Comment (mashu-daishi) [inline schemas/observation/digital-signature.js:21]: Just to confirm: the only **two** people that will be digitally-signing an observation are the observer and the observee, correct?

File: schemas/observation/digital-signature.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+var Joi = require( 'joi' );
+
+var signatureInfo = {
```
Comment (mashu-daishi) [inline schemas/observation/digital-signature.js:5]: Should all three properties be optional? It seems that, if you pass any of them, you should be passing all of them.

File: schemas/observation/digital-signature.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	'sign'    : Joi.string().optional().description( 'string that the user inputs for digital signatures' ),
+	'date'    : Joi.date().optional().description( 'The date of which the user signed the observation' ),
```
Comment (mashu-daishi) [inline schemas/observation/digital-signature.js]: You're probably going to want to type this as `Joi.date().timestamp()...`. We'll want to ensure time is included with the date.

**Review by mashu-daishi — CHANGES_REQUESTED** [review]

File: schemas/observation/digital-signature.js (hunk 3 above)
Comment (Darbiol) [reply schemas/observation/digital-signature.js:21]: yes

File: schemas/observation/digital-signature.js (hunk 2 above)
Comment (Darbiol) [reply schemas/observation/digital-signature.js:17]: yes we are allowing empty arrays

**Comment by Darbiol** [comment]
@mashu-daishi @hanchan07 @MrAnderson0309 i've updated the PR, can you take a look at it again please

File: schemas/observation/digital-signature.js (hunk 2 above)
Comment (mashu-daishi) [reply schemas/observation/digital-signature.js:17]: What's the purpose of the empty array? If it's empty, why not just not save the key?

**Review by mashu-daishi — APPROVED** [review]

File: schemas/observation/digital-signature.js (hunk 2 above)
Comment (Darbiol) [reply schemas/observation/digital-signature.js:17]: so that we can ensure that the output will always be an array

**Pull request merged**